modules/infra/control/port.c

```diff
@@ -461,6 +508,162 @@ static struct iface_type iface_type_port = {
+			p->multicast[i].refcnt++;
+			return 0;
+		} else if (p->multicast[i].refcnt == 0) {
+			memcpy(&p->multicast[i].mac, mcast_addr, sizeof(struct rte_ether_addr));
+			p->multicast[i].refcnt = 1;
+			found = true;
+		}
+	}
+
+	if (i == GR_PORT_NUM_MCAST_ADDR) {
+		p->flags |= GR_PORT_F_ALLMULTI_TOO_MANY_MCAST;
+		LOG(INFO,
+		    "Too many multicast addresses requested for %s",
+		    port_get_iface(p->port_id)->name);
+		goto enable_allmulti;
+	}
+
+	count = mcast_count(p);
+	for (i = 0; i < count; i++) {
+		memcpy(&mac[i], &p->multicast[i].mac, sizeof(struct rte_ether_addr));
+	}
+
+	ret = rte_eth_dev_set_mc_addr_list(p->port_id, mac, count);
+
+	if (ret == 0) {
+		return 0;
+	} else if (ret == -ENOSPC) {
+		LOG(WARNING,
+		    "Enabling mc addr list: no space left on iface %s.",
+		    port_get_iface(p->port_id)->name);
+		p->flags |= GR_PORT_F_ALLMULTI_TOO_MANY_MCAST;
+	} else if (ret == -ENOTSUP) {
+		LOG(WARNING,
+		    "Enabling mc addr list not supported for iface %s",
+		    port_get_iface(p->port_id)->name);
+		p->flags |= GR_PORT_F_ALLMULTI_DEVICE_LIMIT;
+	}
+
+enable_allmulti:
+	return rte_eth_allmulticast_enable(p->port_id);
+}
+
+static int mac_refcnt_qsort(const void *p1, const void *p2) {
+	const struct port_mac_addr *m1 = p1;
+	const struct port_mac_addr *m2 = p2;
+
+	return m2->refcnt - m1->refcnt;
+}
+
+int gr_port_mcast_addr_del(struct iface_info_port *p, struct rte_ether_addr *mcast_addr) {
+	struct rte_ether_addr mac[GR_PORT_NUM_MCAST_ADDR];
+	const struct iface *iface;
+	int i;
+
+	if (p->flags & GR_PORT_F_ALLMULTI_TOO_MANY_MCAST) {
+		// can't disable all multi as we weren't able to keep track of requests;
+		return 0;
+	}
+
+	for (i = 0; i < GR_PORT_NUM_MCAST_ADDR; i++) {
+		if (rte_is_same_ether_addr(mcast_addr, &p->multicast[i].mac)) {
+			p->multicast[i].refcnt--;
+			if (p->multicast[i].refcnt == 0) {
+				memset(&p->multicast[i].mac, 0, sizeof(struct rte_ether_addr));
+				break;
+			} else {
+				return 0;
+			}
+		}
+	}
+
+	if (i == GR_PORT_NUM_MCAST_ADDR) {
+		LOG(INFO,
+		    "multicast addr wasn't allowed for iface %s",
+		    port_get_iface(p->port_id)->name);
+		return -ENODATA;
+	}
+
+	qsort(p->multicast, GR_PORT_NUM_MCAST_ADDR, sizeof(p->multicast[0]), mac_refcnt_qsort);
+
+	p->flags &= ~GR_PORT_F_ALLMULTI_DEVICE_LIMIT;
+	for (i = 0; i < mcast_count(p); i++) {
+		memcpy(&mac[i], &p->multicast[i].mac, sizeof(struct rte_ether_addr));
+	}
+
+	if (mcast_count(p)) {
+		if (rte_eth_dev_set_mc_addr_list(p->port_id, mac, mcast_count(p)) < 0) {
+			LOG(ERR,
+			    "enabling mc addr list failed on iface %s.",
+			    port_get_iface(p->port_id)->name);
+			p->flags |= GR_PORT_F_ALLMULTI_DEVICE_LIMIT;
+			return rte_eth_allmulticast_enable(p->port_id);
+		}
+	} else {
+		rte_eth_dev_set_mc_addr_list(p->port_id, NULL, 0);
+	}
+
+	iface = port_get_iface(p->port_id);
+	if (p->flags & ~GR_PORT_F_ALLMULTI_DEVICE_LIMIT && iface->flags & ~GR_IFACE_F_ALLMULTI) {
+		return rte_eth_allmulticast_disable(p->port_id);
+	}
```

Comment:
No braces please. And maybe this test can be shortened:

```c
if (p->flags & ~(GR_PORT_F_ALLMULTI_DEVICE_LIMIT | GR_IFACE_F_ALLMULTI))
```

Reply:
Can't be shortened as  port flags and iface flags are different.